### PR TITLE
ENH add ensemble notebook

### PR DESCRIPTION
**Pull request by glemaitre**
Add a notebook to present the principle of bagging and additional randomization to have random-forest.
Add a presentation of the gradient-boosting.
Add a section regarding the important parameter and expected values of the random forest and gradient boosting.

### CODE REVIEW

**Comment by glemaitre** [comment]
OK I think that we have a good draft here. @ogrisel @lesteve @GaelVaroquaux @TwsThomas

**Comment by glemaitre** [comment]
Only the last part is too verbose and we should add more code part most probably with a grid-search. I will do that probably tomorrow.

File: python_scripts/ensemble.py [hunk 1]
```diff
@@ -0,0 +1,972 @@
...
+
+# %% [markdown]
+# We can observe that there is several samples for which the current classifier
+# was not able to make the proper decision. As we previously mentioned,
+# boosting relies on creating a new classifier which will try to correct these
+# error. In scikit-learn, learners usually support a parameter `sample_weight`
+# which allows to specify to pay more attention to some specific samples. This
+# parameters is set when calling `classifier.fit(X, y, sample_weight=weights)`.
+# We will use this trick to create a new classifier by discarding all well
+# classified samples and only consider the misclassified samples. Thus,
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:557]:
```suggestion
# We will use this trick to create a new classifier by discarding all the
# already well classified samples and thus only considering the misclassified samples. Thus,
```

File: python_scripts/ensemble.py [hunk 2]
```diff
@@ -0,0 +1,972 @@
...
+# classifier which will be different by focusing more or less to some specific
+# samples of the dataset. Then, boosting is different from bagging: here we
+# never resample our dataset, we just assign weight to the original dataset.
+#
+# Finally, we see that we need some strategy to combine the learners together:
+# * one need to define a way to compute the weight which need to be assigned to
+#   samples;
+# * one need to assign a weight to each learner when making the predictions.
+#
+# Here, we defined really simple scheme to define the weight. However, there is
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:620]:
```suggestion
# Here, we defined a really simple scheme to assign the weight. However, there is
```

File: python_scripts/ensemble.py [hunk 3]
```diff
@@ -0,0 +1,972 @@
...
+# some statistical basis which can be used to theoretically defined these
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:621]:
```suggestion
# some statistical basis which can be used to theoretically assigned these
```

File: python_scripts/ensemble.py [hunk 4]
```diff
@@ -0,0 +1,972 @@
...
+        [x[idx], x[idx]], [y[idx], y_pred_raw[idx]], color="red",
+    )
+
+_ = plt.legend(
+    [line_predictions[0], lines_residuals[0]], ["Fitted tree", "Residuals"]
+)
+
+# %% [markdown]
+# Since the tree is underfitting the data, the accuracy of the tree will is far
+# to be perfect on the training data. We can observe it on the figure by
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:699]:
```suggestion
# Since the tree is underfitting the data, its accuracy will be far from
# perfect on the training data. We can observe it on the figure by
```

File: python_scripts/ensemble.py [hunk 5]
```diff
@@ -0,0 +1,972 @@
...
+    lines_residuals = plt.plot(
+        [x[idx], x[idx]], [residuals[idx], y_pred_residuals[idx]], color="red",
+    )
+
+_ = plt.legend(
+    [line_predictions[0], lines_residuals[0]], ["Fitted tree", "Residuals"]
+)
+
+# %% [markdown]
+# We see that this new tree succeed to correct the residual for some specific
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:735]:
```suggestion
# We see that this new tree only succeed to correct the residual of some specific
```

File: python_scripts/ensemble.py [hunk 6]
```diff
@@ -0,0 +1,972 @@
...
+)
+print(f"Error of the tree: {y_true - y_pred_first_and_second_tree:.3f}")
+
+# %% [markdown]
+# We choose a sample for which only 2 trees were enough to make the perfect
+# prediction. However, we saw in the previous plot that 2 trees were not enough
+# to correct the residuals for all samples. In this regard, one need to add
+# several trees in the ensemble to succeed to correct the error.
+#
+# We will make a small comparison between random-forest and gradient boosting
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:815]:
```suggestion
# We will make a performance comparison between random-forest and gradient boosting
```

File: python_scripts/ensemble.py [hunk 7]
```diff
@@ -0,0 +1,972 @@
...
+# We can observe that in our grid-search, the larger `max_depth` with the
+# largest `n_estimators` lead to the best performance.
+#
+# Gradient-boosting decision tree
+#
+# In gradient-boosting, parameters tuning is a combination of several
+# parameters: `n_estimators`, `max_depth`, and `learning_rate`.
+#
+# Let's first discuss about the `max_depth`. We saw in the section of the
+# gradient-boosting that the algorithm fit the error of the previous trees in
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:896]:
```suggestion
# Let's first discuss about the `max_depth`. We saw that
# gradient-boosting algorithm fit the error of the previous trees in
```

File: python_scripts/ensemble.py [hunk 8]
```diff
@@ -0,0 +1,972 @@
...
+# the ensemble. Thus, fitting fully grown trees will be detrimental. Indeed,
+# the first tree of the ensemble will overfit the data and thus no subsequent
+# tree is required since there will be no residuals to be fitted. Therefore,
+# the tree used in gradient-boosting will have a low depth between 3 to 8
+# levels typically.
+#
+# Having in mind this consideration, deeper will be the trees, the residuals
+# will be corrected faster and with a lower number of estimators. So
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:904]:
```suggestion
# Having in mind this consideration, deeper will be the trees, faster the residuals
# will be corrected and with a lower number of estimators. So
```

File: python_scripts/ensemble.py [hunk 9]
```diff
@@ -0,0 +1,972 @@
...
+# `n_estimators` should be increased if `max_depth` is lower.
+#
+# Finally, we did not explain what `learning_rate` parameter was corresponding
+# too. When fitting the residual one could choose if the tree should try to
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:908]:
```suggestion
# to. When fitting the residual one could choose if the tree should try to
```

File: python_scripts/ensemble.py [hunk 10]
```diff
@@ -0,0 +1,972 @@
...
+# correct all possible errors or only a fraction of it. The learning-rate is
+# allowing to control this behaviour. A small value of learning-rate will only
+# correct the residuals of very few samples. If the learning-rate is set to 1
+# then we will find a tree that fit the residuals of all samples. So, with a
+# very low learning-rate, we will need more estimators to correct the error.
+# However, a too large learning-rate will tend obtain an overfitted ensemble
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:914]:
```suggestion
# However, a too large learning-rate will tend to obtain an overfitted ensemble
```

File: python_scripts/ensemble.py [hunk 11]
```diff
@@ -0,0 +1,972 @@
...
+# similarly to have a too large tree depth.
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:915]:
```suggestion
# similarly to having a too large tree depth.
```

File: python_scripts/ensemble.py [hunk 12]
```diff
@@ -0,0 +1,972 @@
...
+grid_search.fit(X_train, y_train)
+
+cv_results = pd.DataFrame(grid_search.cv_results_)
+cv_results[columns].sort_values(by="rank_test_score")
+
+# %% [markdown]
+# ## Accelerate gradient-boosting
+#
+# We previously mentioned that random-forest is an efficient algorithm since
+# each tree of the ensemble can be fitted at the same time from an independent
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:935]:
```suggestion
# each tree of the ensemble can be fitted at the same time in an independent
```

File: python_scripts/ensemble.py [hunk 13]
```diff
@@ -0,0 +1,972 @@
...
+# manner.
+#
+# In gradient-boosting, the algorithm is a sequential algorithm. it requires
+# the `N-1` trees to fit the tree at the stage `N`. Therefore, the algorithm
+# is quite computationally expensive. The most expensive in this algorithm is
```
Comment (TwsThomas) [inline python_scripts/ensemble.py:940]:
```suggestion
# is quite computationally expensive. The most expensive part in this algorithm is
```

**Comment by glemaitre** [comment]
Thanks @TwsThomas I will make the changes (if they applied since I did a second round of review :))
